gh-74696: Pass root_dir to custom archivers which support it

serhiy-storchaka opened the pull request:
<!-- gh-issue-number: gh-74696 -->
* Issue: gh-74696
<!-- /gh-issue-number -->

serhiy-storchaka committed:
gh-74696: Pass root_dir to custom archivers which support it

merwok commented on Doc/library/shutil.rst:
```diff
@@ -575,7 +575,8 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
 
       This function is not thread-safe when custom archivers registered
-      with :func:`register_archive_format` are used.  In this case it
+      with :func:`register_archive_format` does not support the *root_dir*
```
```suggestion
      with :func:`register_archive_format` do not support the *root_dir*
```

(subject is _custom archivers_)

serhiy-storchaka committed:
Update Doc/library/shutil.rst

Co-authored-by: Éric <merwok@netwok.org>

Commit:
Merge branch 'main' into shutil-register_archive_format-supports_root_dir-attr

NoSuck committed:
closes gh-97650: correct sphinx executable (gh-97651)

jeff5 committed:
gh-96397: Document that attributes need not be identifiers (#96454)

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

ofey404 committed:
gh-96348: Deprecate the 3-arg signature of coroutine.throw and generator.throw (GH-96428)

ehebert committed:
Use SyntaxError invalid range in tutorial introduction example (GH-93031)



Use output from a 3.10+ REPL, showing invalid range, for the
SyntaxError examples in the tutorial introduction page.

Automerge-Triggered-By: GH:iritkatriel

zooba committed:
gh-97649: The Tools directory is no longer installed on Windows (GH-97653)

zooba committed:
gh-90989: Install Windows launcher per-user, and clarify some installer text (GH-97655)

vstinner committed:
gh-94526: getpath_dirname() no longer encodes the path (#97645)

Fix the Python path configuration used to initialized sys.path at
Python startup. Paths are no longer encoded to UTF-8/strict to avoid
encoding errors if it contains surrogate characters (bytes paths are
decoded with the surrogateescape error handler).

getpath_basename() and getpath_dirname() functions no longer encode
the path to UTF-8/strict, but work directly on Unicode strings. These
functions now use PyUnicode_FindChar() and PyUnicode_Substring() on
the Unicode path, rather than strrchr() on the encoded bytes string.

csabella committed:
bpo-35675: IDLE  - separate config_key window and frame (#11427)

bpo-35598: IDLE: Refactor window and frame class

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

gpshead committed:
gh-87597: Document TimeoutExpired.stdout & .stderr types (#97685)

This documents the behavior that has always been the case since timeout
support was introduced in Python 3.3.

gvanrossum committed:
GH-96827: Don't touch closed loops from executor threads (#96837)

* When chaining futures, skip callback if loop closed.
* When shutting down an executor, don't wake a closed loop.

gvanrossum committed:
GH-97592: Fix crash in C remove_done_callback due to evil code (#97660)

Evil code could cause fut_callbacks to be cleared when PyObject_RichCompareBool is called.

ericsnowcurrently committed:
gh-90110: Update the c-analyzer Tool (gh-97695)

https://github.com/python/cpython/issues/90110

gpshead commented on Lib/test/test_shutil.py:
```diff
@@ -1570,26 +1570,62 @@ def test_tarfile_root_owner(self):
-        def _breaks(*args, **kw):
+        def archiver(base_name, base_dir, **kw):
+            self.assertNotIn('root_dir', kw)
+            if root_dir is None:
```
This is referring to nonlocal variable `root_dir` which is first "defined" by being set far below. That makes for confusing code. Rather than an `if root_dir ...:` conditional within the archiver, it would be cleaner to register and unregister a different archiver for each of the two test scenarios below.

ambv committed:
gh-90908: Document asyncio.Task.cancelling() and asyncio.Task.uncancel() (#95253)

Co-authored-by: Thomas Grainger <tagrain@gmail.com>

hawkinsw committed:
Fix capitalization of Unix in documentation (#96913)

gpshead committed:
gh-95588: Drop the safety claim from `ast.literal_eval` docs. (#95919)

It was never really safe and this claim conflicts directly with the big warning in the docs about it being able to crash the interpreter.

ofey404 committed:
gh-97591: In `Exception.__setstate__()` acquire strong references before calling `tp_hash` slot (#97700)

CAM-Gerlach committed:
gh-95975: Move except/*/finally ref labels to more precise locations (#95976)

* gh-95975: Move except/*/finally ref labels to more precise locations

* Add section headers to fix :keyword: role and aid navigation

* Move see also to the introduction rather than a particular subsection

* Fix other minor Sphinx syntax issues with except

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

* Suppress redundant link to same section for except too

* Don't link try/except/else/finally keywords if in the same section

* Format try/except/finally as keywords in modified sections

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

CAM-Gerlach committed:
gh-97607: Fix content parsing in the impl-detail reST directive (#97652)

* Don't parse content as arg in the impl-detail directive

This does not change the (untranslated) output,
but ensures that the doctree node metadata is correct.
which fixes gh-97607 with the text not being translated.
It also simplifies the code and logic
and makes it consistant with the docutils built-in directives.

* Remove unused branch from impl-detail directive handling no-content case

This is not used anywhere in the docs and lacks a clear use case,
and is more likely a mistake which is now flagged at build time.
This simplifies the logic from two code paths to one,
and makes the behavior consistant with similar built-in directives
(e.g. the various admonition types).

* Further simplify impl-detail reST directive code

vsajip committed:
[docs] Update logging cookbook with recipe for using a logger like an output… (GH-97730)

serhiy-storchaka committed:
Refactor tests.

serhiy-storchaka committed:
Merge branch 'main' into shutil-register_archive_format-supports_root_dir-attr

gvanrossum commented on Doc/library/shutil.rst:
```diff
@@ -614,12 +615,21 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
+   If *function* has the *supports_root_dir* attribute set to ``True``,
+   the *root_dir* argument is passed as a keyword argument.
+   Otherwise the current working directory of the process is temporarily
+   changed before calling *function*.
```
Maybe state to what it is changed? (IIUC it does `os.chdir(root_dir)`.)

merwok replied:
```suggestion
   changed to *root_dir* before calling *function*.
```

gvanrossum commented on Doc/library/shutil.rst:
```diff
@@ -614,12 +615,21 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
    Further arguments are passed as keyword arguments: *owner*, *group*,
    *dry_run* and *logger* (as passed in :func:`make_archive`).
 
+   If *function* has the *supports_root_dir* attribute set to ``True``,
```
It wasn't until I read the code before I understood what this line means. It's something like `getattr(function, "supports_root_dir", False)`. I wonder if you could clarify that somehow?

merwok replied:
Agreed, as it’s not a regular attribute of all functions.  Mayge:

```suggestion
   If *function* has the custom attribute `function.supports_root_dir` set to ``True``,
```

gvanrossum reviewed (approved):
Just two doc nits, entirely optional.

merwok commented on Doc/library/shutil.rst:
```diff
@@ -575,7 +575,8 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
+      with :func:`register_archive_format` do not support the *root_dir*
+      argument.  In this case it
       temporarily changes the current working directory of the process
       to perform archiving.
```
```suggestion
      to *root_dir* to perform archiving.
```

merwok commented on Doc/library/shutil.rst:
```diff
@@ -614,12 +615,21 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
    list of archivers.  Defaults to an empty string.
 
+   .. versionchanged:: 3.12
+      Added support of functions supporting the *root_dir* argument.
```
```suggestion
      Added support for functions supporting the *root_dir* argument.
```

serhiy-storchaka committed:
Apply suggestions from code review

Co-authored-by: Éric <merwok@netwok.org>

merwok commented on Doc/library/shutil.rst:
```diff
@@ -614,12 +615,21 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
    Further arguments are passed as keyword arguments: *owner*, *group*,
    *dry_run* and *logger* (as passed in :func:`make_archive`).
 
+   If *function* has the custom attribute `function.supports_root_dir` set to ``True``,
```
```suggestion
   If *function* has the custom attribute ``function.supports_root_dir`` set to ``True``,
```

merwok committed:
fix markup

gvanrossum reviewed (approved):
Excellent. @serhiy-storchaka feel free to land!

serhiy-storchaka commented on Doc/whatsnew/3.12.rst:
```diff
@@ -127,6 +127,15 @@ os
+* :func:`shutil.make_archive` now passes the *root_dir* argument to custom
+  archivers which support it.
+  In this case it no longer temporarily changes the current working directory
+  of the process to perform archiving.
```
```suggestion
  of the process to *root_dir* to perform archiving.
```

serhiy-storchaka committed:
Update Doc/whatsnew/3.12.rst
